File: algorithms/tokenBucket.js
const { getRedisClient } = require("../redis/redisCluster");

const CAPACITY = 10;
const REFILL_RATE = 1;

async function tokenBucket(key) {

  const redis = getRedisClient(key); // ✅ FIX

  const bucket = await redis.hmget(key, "tokens", "lastRefill");

  let tokens = bucket[0] ? parseFloat(bucket[0]) : CAPACITY;
  let lastRefill = bucket[1] ? parseInt(bucket[1]) : Date.now();

  const now = Date.now();
  const elapsed = (now - lastRefill) / 1000;

  const refill = elapsed * REFILL_RATE;
  tokens = Math.min(CAPACITY, tokens + refill);

  if (tokens < 1) {
    return false;
  }

  tokens -= 1;

  await redis.hmset(key, {
    tokens,
    lastRefill: now
  });

  await redis.expire(key, 120);

  return true;
}

module.exports = tokenBucket;

File: lua/tokenBucket.lua
local key = KEYS[1]

local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

local bucket = redis.call("HMGET", key, "tokens", "lastRefill")

local tokens = tonumber(bucket[1])
local lastRefill = tonumber(bucket[2])

if tokens == nil then
  tokens = capacity
  lastRefill = now
end

local delta = math.max(0, now - lastRefill)
local refill = delta * refill_rate / 1000

tokens = math.min(capacity, tokens + refill)

if tokens < 1 then
  return 0
end

tokens = tokens - 1

redis.call("HMSET", key, "tokens", tokens, "lastRefill", now)
redis.call("EXPIRE", key, 120)

return 1